pop-api/examples/fungibles/lib.rs

```diff
@@ -1,108 +1,268 @@
 		#[ink(message)]
-		pub fn allowance(
-			&self,
-			token: TokenId,
-			owner: AccountId,
-			spender: AccountId,
-		) -> Result<Balance> {
-			api::allowance(token, owner, spender)
+		fn balance_of(&self, owner: AccountId) -> Balance {
+			api::balance_of(self.id, owner).unwrap_or_default()
 		}
 
+		/// Returns the allowance for a `spender` approved by an `owner`.
+		///
+		/// # Parameters
+		/// - `owner` - The account that owns the tokens.
+		/// - `spender` - The account that is allowed to spend the tokens.
 		#[ink(message)]
-		pub fn transfer(&mut self, token: TokenId, to: AccountId, value: Balance) -> Result<()> {
-			api::transfer(token, to, value)
+		fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+			api::allowance(self.id, owner, spender).unwrap_or_default()
 		}
 
+		/// Transfers `value` amount of tokens from the caller's account to account `to`
+		/// with additional `data` in unspecified format.
+		///
+		/// # Parameters
+		/// - `to` - The recipient account.
+		/// - `value` - The number of tokens to transfer.
+		/// - `data` - Additional data in unspecified format.
 		#[ink(message)]
-		pub fn transfer_from(
+		fn transfer(
 			&mut self,
-			token: TokenId,
-			from: AccountId,
 			to: AccountId,
 			value: Balance,
 			_data: Vec<u8>,
-		) -> Result<()> {
-			api::transfer_from(token, from, to, value)
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `to` is the same address or `value` is zero.
+			if caller == to || value == 0 {
+				return Ok(());
+			}
+			api::transfer(self.id, to, value).map_err(Psp22Error::from)?;
+			self.env().emit_event(Transfer { from: Some(caller), to: Some(to), value });
+			Ok(())
 		}
 
+		/// Transfers `value` tokens on behalf of `from` to the account `to`
+		/// with additional `data` in unspecified format.
+		///
+		/// # Parameters
+		/// - `from` - The account from which the token balance will be withdrawn.
+		/// - `to` - The recipient account.
+		/// - `value` - The number of tokens to transfer.
+		/// - `data` - Additional data with unspecified format.
 		#[ink(message)]
-		pub fn approve(
+		fn transfer_from(
 			&mut self,
-			token: TokenId,
-			spender: AccountId,
+			from: AccountId,
+			to: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::approve(token, spender, value)
+			_data: Vec<u8>,
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if `from` and `to` is the same address or `value` is zero.
+			if from == to || value == 0 {
+				return Ok(());
+			}
+			// If `from` and the caller are different addresses, a successful transfer results
+			// in decreased allowance by `from` to the caller and an `Approval` event with
+			// the new allowance amount is emitted.
+			api::transfer_from(self.id, from, to, value).map_err(Psp22Error::from)?;
+			self.env().emit_event(Transfer { from: Some(caller), to: Some(to), value });
```

Comment:
```suggestion
			self.env().emit_event(Transfer { from: Some(from), to: Some(to), value });
```

In this example, the caller is just a proxy and the use case is relying on the fact that from has pre-approved the contract to manage its funds - e.g. smart contract waller.